cilium/CFP-32820-improve-scalability-of-health-probing.md

```diff
@@ -0,0 +1,158 @@
+
+A possible solution for this may be to continue refreshing the connectivity cache at intervals of 60 seconds. Instead of storing the start time for the health report as a whole, we may store the health report start time on a per-node basis. Then, every 60 seconds, we make a call to refresh the cache, and update metrics for nodes iff the start time for the node’s last stored metric is before the start time of the current health report.
+
+#### Metrics Update Period
+
+As a consequence of this proposal, health probe updates may take longer and be more unpredictably-timed for users. In order to keep users aware of how long it will take to receive accurate metrics for the known nodes in the cluster, we can update the output of `cilium-health status` to also report the current probe interval and/or the most recent timestamp at which metrics for all known nodes were updated.
+
+#### Probe Timeouts and Errors
+
+The HTTP probe is currently set to have a timeout of 30 seconds. The ICMP probe has MaxRTT set to 60 seconds, which acts as the timeout for the pings. Both timeouts seem larger than necessary. The HTTP probe timeout can be re-configured to 10 seconds, which is a reasonable timeout and accounts for the sizes of the maximum bucket configured for the histogram metric `node_health_connectivity_latency_seconds`.
+
+However, for the ICMP probe, by nature of the library `servak/go-fastping`, the probe interval value and max RTT cannot easily be separated. This is one reason to consider changing to the library `prometheus-community/pro-bing`, which will allow us to easily configure different Timeout and Interval values, allowing us to set the timeout to 10 seconds and the interval to the probe interval value.
+
+#### Synchronous Probe Behavior
+Currently, a synchronous probe is run in two situations:
+1. When the health server starts running
+2. When an HTTP GET request is made to the health server for /status/probe, usually by a call to `cilium-health status --probe`.
+
+We will plan to deprecate the capability to synchronously probe the health server via the `cilium-health` CLI command. This will reduce the negative effects of synchronous probing on resource usage at high-scale. On start-up, probing will rely on the probe interval determined by cluster size, as in the typical case.
+
+## Key Questions
+
+### Key Question:
+
+What is the preferable approach to configuring the QPS?
+
+### Option 1:
+
+Configure the flag `--connectivity-probe-frequency-ratio` described in this proposal.
+
+#### Pros
+
+* This flag abstracts away decision-making regarding the base interval from the user, and instead allows the user to make a decision regarding the trade-off between probe frequency and resource usage with a well-defined range of possible values and ease-of-understanding of a ratio between 0 and 1.
+
+#### Cons
+
+* The obfuscation may be more complex to interpret and document.
+* Removes some direct configuration power from user.
+
+### Option 2:
```

Comment:
```suggestion
### Option 2 (Discarded):
```

Reply:
Maybe there's a better way to mark this but I think it makes sense to capture that these other options were considered but it seems like they're not the direction we want to go.